### DISCUSSION TITLE
Add r2u repository

### DISCUSSION THREAD

**Discussion by travis-82** [dat-linux/community#15] [App, Extra, Book suggestions]
Hi and many thanks for your great work.

There is a binary repository of R packages for Ubuntu LTS called r2u. It would be nice and helpful if you add this repository and configure R so that it grab packages from this repository by default. Please have look:

https://github.com/eddelbuettel/r2u

**Comment by dat-linux** [comment]
Thanks, this is something that could come as a configurable "extra". In the meantime feel free to set up this repo and use it on your own, I can't see why it wouldn't be compatible with DAT Linux. It can take over the "install.packages()" calls in R, so this is something to be mindful of. One other thing to consider is how the repos handle conflicting packages (the default Ubuntu package repo has some prebuilt ones already). And how up to date are the packages, how quickly do they get built when a package is updated on CRAN?

> By choice, it does not map the remove.packages() for package removal.. Packages can be uninstalled via the system package manager 

I think this is a bit of a hurdle, as it requires apt to remove packages which you'd normally easily do in R or RStudio. 

The thing about the default r-cran-* packages in Ubuntu is that if they're outdated then the R package manager can update them from CRAN, so you have the latest version. If you override the R package installer then it solely relies on the new repo as being up to date. 

Further, what happens if/when you choose to drop this repo and revert back? 

There's a number of issues that need to be considered if trying this out, and relies on more advanced skills, which leads me to think this should be either assisted (adding a utility layer to ease the process) or not included by default.

**Comment by travis-82** [comment]
Thanks for the answer and your points are valid. However I think there are workarounds:
R2u packages are built against latest R version that is the default version in Dat-linux, Hence, adding this repository to Ubuntu is same as adding a third-party ppa for newer versions of desired apps. Taking over "install.packages()" calls in R is not mandatory and it depends on installing and enabling bspm. Without bspm, users either can use apt to install r-cran packages system-wide from r2u repository or use install.packages() to build and install cran packages from source to their home directory. 
compiling R packages is a time consuming and bug prone process. I have friends who ditched Linux because of that. Projects like r2u for Ubuntu and iucar/cran copr for fedora would be very helpful for this type of users.

**Reply by dat-linux** [reply]
It's just not really a user friendly proposition.

Right now, calling install.packages() will get the latest package from CRAN, regardless if the system package already has a previous/old version. The APT packages (via the default Ubuntu repo) are installed as a convenience, to ensure a base R system is there. Updates via CRAN will detect an update and install it in the user environment, it's seamless. Even if someone installs a package via apt-get then R will pick it up and make it available. They complement each other.

Not using this feature via install.packages() defeats the purpose, most people just want to be able to get the latest package, and usually from RStudio. Which means they're using install.packages() either explicitly or behind the scenes. Again, it just makes the apt repo irrelevant.

Sure, this helps provide pre-built binaries, but the cost to achieve it is a bit high imo, especially given the extra steps you need to take, plus the point I made re rolling back. It's potentially risky and could leave a system unstable (also potential for conflicting packages between repos). There's also no guarantee the prebuilt packages are failsafe, or that it will be around for long - projects do get zapped.

Again, there's nothing stopping anyone from taking these advanced steps if they wish, and understand the risks involved. But I don't think it should be the default setup. It's Linux, unlike Windows or Macs you can do anything you want to set it up as you wish.

I've never had problems with Linux and R packages being built, beyond occasional lengthy compiles, but R has been smart enough to know if it failed/succeeded. Very rarely is it due to a missing shared library - but these problems are almost always documented somewhere, just google it. Not sure how this would be different for Windows or Mac -does CRAN pre-build them?

And I still don't know how long it takes them to build a newer version of a package and make it available. Is it a rolling release that updates packages as they are updated on CRAN? If they're behind CRAN on releases by days then the CRAN package manager is going to have them available sooner anyway. 

You also should understand that base R packages are installed and updated via the Ubuntu apt repo. This works pretty well so far, and it ensures the base packages are there at all times - ones that a user will typically not install via CRAN. There's a risk that mixing repos might affect this stability. Sure, a user can do something to affect this, but I repeat this kind of added risk I don't think should be the **default** setup.

**Comment by dat-linux** [comment]
It's all moot anyway, as imposing such a change through a DAT Linux system update (which is focused on the control panel) isn't the right thing to do, and if at all it would have to be in an ISO upgrade release - which at this stage this is not planned until 24.04.

Keeping it on the back burner. At best it can be an added extra with some guidance via install scripts, etc. But not a priority until/if it's more thoroughly tested.
